### DISCUSSION TITLE
IndexedDB not persisted

### DISCUSSION THREAD

**Discussion by mgscreativa** [pmndrs/zustand#1093] [Q&A] [answered]
Hi! I'm having a hard time setting up IndexedDB with idb-keyval and zustand. First of all, I'm using Expo to develop a custom RN app using expo development builds.

I can use zustand without any issue to setup fishes state and **addFish** to add plus one fish each time it's called, but the persisted storage that I want to use with IndexedDB doesn't seem to get persisted at all, because, after reload or reboot of the app, the storage gets value 0 on the **fishes** key. 

**My deps are**
```
    "idb-keyval": "^6.2.0",
    "zustand": "^3.7.2"
```

**usePersistAppContextStore.tsx**
```
import create from 'zustand';
import { persist, StateStorage } from 'zustand/middleware';
import { get, set, del } from 'idb-keyval';

const zustandStorage: StateStorage = {
  getItem: async (name: string): Promise<string | null> => {
    return (await get(name)) || null;
  },
  setItem: async (name: string, value: string): Promise<void> => {
    await set(name, value);
  },
  removeItem: async (name: string): Promise<void> => {
    await del(name);
  },
};

const usePersistAppContextStore = create(
  persist(
    (set, get) => ({
      fishes: 0,
      addFish: () => set({ fishes: get().fishes + 1 }),
      clearFish: () => set({ fishes: 0 }),
    }),
    {
      name: 'app-storage',
      getStorage: () => zustandStorage,
    },
  ),
);

export default usePersistAppContextStore;
```

**App.tsx**
```
...
...
import usePersistAppContextStore from './app/stores/usePersistAppContextStore';

export default function App() {
  const { fishes, addFish } = usePersistAppContextStore(({ fishes, addFish }) => ({
    fishes,
    addFish,
  }));

  console.log('Stored fishes:', fishes);
  addFish();
  console.log('Added one fish, equals to ', fishes, ' fishes');
...
...
```

The code above add fishes but after restart of reboot app, the counter starts at 0 again.

**Comment by dai-shi** [comment]
Can anyone help this?

**Comment by AnatoleLucet** [comment]
Seems to be a dup of #562.
You need to wait until the hydration is done before mutating the store:
```ts
export default function App() {
  const { addFish } = usePersistAppContextStore();

  useEffect(() => {
    usePersistAppContextStore.persist.onFinishHydration((s) => {
      console.log("Stored fishes:", s);
      addFish();
      console.log("Added one fish, equals to ", s.fishes + 1, " fishes");
    });
  }, []);
```

See [this part of the doc](https://github.com/pmndrs/zustand/blob/main/docs/persisting-store-data.md#how-can-i-check-if-my-store-has-been-hydrated) for potentially more suiting ways to await hydration.

**Comment by mgscreativa** [comment]
Hi @AnatoleLucet and @dai-shi ! Thanks for the replies. I have already read the docs, and as the rehydration didn't work as expected, I decided to fill this question here. I followed the docs with the **onRehydrateStorage** callback and the **useHydration hook**, none of them worked

I have took the time to compile a little Expo snack to reproduce the issue, it basically adds zustand and idb-keyval deps but I think that the hydration process fails somehow, because the useHydration hook never returns true. The example can be run on any phone with the Expo Go app https://snack.expo.dev/@mgscreativa/zustand-indexeddb-hydration-issue

Here is a GH repo for testing https://github.com/mgscreativa/zustand-idb-keyval-persist-issue

On my phone I'm getting this warning
```
[Unhandled promise rejection: TypeError: undefined is not an object (evaluating 'state.setHasHydrated')]
at stores/usePersistAppContextStore.js:34:13 in <anonymous>
at node_modules/zustand/middleware.js:421:74 in toThenable.then...then._catch$argument_0
```

**Edit:**
Added some console logs and I get this
```
app-storage has been retrieved
hasHydrated: false
app-storage with value {"state":{"fishes":0,"hasHydrated":true},"version":0} has been saved
iOS Running app on iPhone XS Max
hydrated: false
```

**Reply by AnatoleLucet** [reply]
Ah my bad, i thought the issue was happening on a web app. IndexedDB doesn't seem to be supported by React Native.
You might want to use React Native's `AsyncStorage`.

FYI, you can see the "real" error by doing the following:
```ts
      onRehydrateStorage: () => (state, err) => {
        if (err) {
          console.error(err)
          return
        }
        
        state.setHasHydrated(true)
      }
```
I'll see if i can make a few changes to enforce error checking for easier debbuging.

**Reply by mgscreativa** [reply]
Yup, this get the real error 

(0, _idbKeyval.get) is not a function. (In '(0, _idbKeyval.get)(name)', '(0, _idbKeyval.get)' is undefined)
at stores/usePersistAppContextStore.js:11:18 in zustandStorage.getItem
at stores/usePersistAppContextStore.js:8:11 in zustandStorage.getItem
at node_modules/zustand/middleware.js:270:21 in
at node_modules/zustand/middleware.js:392:13 in hydrate
at node_modules/zustand/middleware.js:457:11 in
at node_modules/zustand/index.js:75:21 in createStore
at node_modules/zustand/index.js:83:59 in create
at stores/usePersistAppContextStore.js:28:41 in
at node_modules/metro-runtime/src/polyfills/require.js:349:11 in loadModuleImplementation
at App.js:11:0 in
at node_modules/metro-runtime/src/polyfills/require.js:349:11 in loadModuleImplementation
at node_modules/expo/AppEntry.js:3:0 in
at node_modules/metro-runtime/src/polyfills/require.js:349:11 in loadModuleImplementation
at node_modules/metro-runtime/src/polyfills/require.js:201:44 in guardedLoadModule
at http://10.0.0.2:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false&strict=false&minify=false:143231:3 in global code

**Comment by mgscreativa** [comment, marked as answer]
Thanks guys! seems that I need to check another storage solution. Tried MMKV for RN in Expo, works just fine, but I lost all JS debugging because it's JSI.

Thanks anyway!
